SSVEP Paradigm and SSVEPExo dataset

sylvchev opened the pull request:
This PR add the support for SSVEP paradigm, which defines the stimulation frequencies, the number of classes to consider. The returned data is a concatenation of the EEG bandpass filtered around the stim freq (`[f_n-0.5, f_n+0.5]`) and the EEG with a wide bandpass (`[1, 45]`). So for a dataset with 10 electrodes, T time samples and 2 stim freq, the returned data of shape (10*(2+1), T). The first ten rows are the 10 electrodes filtered around the first stim freq, the next 10 rows are filtered around the second freq and the last 10 rows are the EEG filtered with a wide bandpass. There is a unit test for SSVEP paradigm in `moabb/tests/paradigms.py`

A first SSVEP dataset, `SSVEPExo()` is included, which is tested in `examples/plot_cross_subject_ssvep.py`. The dataset have 4 classes (3 stim frequency and a resting class), 2 classes corresponding to the first stim freq are used in the example. I wrote a Canonical Correlation Analaysis (CCA)-based classifier, as it is commonly used in SSVEP. The CCA classifier use the broadband filtered signal. The dataset is also benchmark against Riemannian tangent classifier using extended covariance matrices. These matrices are computed from the concatenation of signal filtered around the stim freq. 

For compatibility issue, I copied the modifications made in the  #51 on `moabb/datasets/fake.py`

vinay-jayaram commented on moabb/paradigms/ssvep.py:
```diff
@@ -0,0 +1,185 @@
+            if not overlap >= self.n_classes:
+                ret = False
+
+        # we should verify list of channels, somehow
```
code to do that exists in dataset.utils, it's done in the dataset_search function

sylvchev commented on examples/plot_cross_subject_ssvep.py:
```diff
@@ -0,0 +1,202 @@
+pipelines['RG + LogReg'] = make_pipeline(
+    FilteredSignal(),
+    Covariances(estimator='lwf'),
+    TangentSpace(),
```
0.5-1 second/subject, on a small Mac Book Pro with Intel i5 proc.

sylvchev commented on examples/plot_cross_subject_ssvep.py:
```diff
@@ -0,0 +1,202 @@
+paradigm = BaseSSVEP(n_classes=2)
+SSVEPExo().download(update_path=True, verbose=False)
+datasets = [SSVEPExo()]
+datasets[0].subject_list = datasets[0].subject_list[:10]
```
ok, done

sylvchev commented on examples/plot_cross_subject_ssvep.py:
```diff
@@ -0,0 +1,202 @@
+# Loading the SSVEP paradigm and the SSVEP Exo dataset, restricting to
+# the first two classes (here 13 and 17 Hz) and the first 10 subjects.
+paradigm = BaseSSVEP(n_classes=2)
+SSVEPExo().download(update_path=True, verbose=False)
```
Yes, only the required files are downloaded now.

sylvchev commented on moabb/paradigms/ssvep.py:
```diff
@@ -0,0 +1,185 @@
+    freqs: list of str | None (default None)
+        List of stimulation frequencies (float). If None, use all stimulus
+        found in the dataset.
+    n_classes: int,
```
Yes, it is optional. It's corrected in the docstring.

sylvchev commented:
I will address the last two comments, I have some ideas to write a properly modified code. I think that it is possible to use only the term "event" instead of "freq" in SSVEP paradigm.
I'll update the PR today or tomorrow.

sylvchev commented:
I rewrote the SSVEP paradigm classes to stick to an approach closer to the one of Alex, distinguishing between SSVEP (single bandpass) and FilterBankSSVEP (useful for extended covariance matrices and RG). I updated the example to reflect the changes. 
I have some ideas to merge the `process_raw` method in the `BaseParadigm` class. I will try and push a commit tonight if possible. We could consider merging after this last item.

sylvchev commented:
I pushed the code of `process_raw` in the `BaseParadigm`, and update SSVEP and MotorImagery classes. You could do the same for P300 paradigm. I added an empty method `prepare_process` in BaseParadigm, which is overloaded in SSVEP to compute the filter base on the chosen dataset. Indeed, addding an empty method overloaded in some subclasses is criticizable, and a solution involving an interface and more advanced OOP is possible. This proposition a more readable and straightforward solution, but I could rewrite it if you want.
I also improved the coverage of tests in `tests/paradigm.py` and cleaned a little bit.

alexandrebarachant commented on examples/plot_cross_subject_ssvep.py:
```diff
@@ -0,0 +1,209 @@
+# extended covariances matrices
+
+
+class FilteredSignal(BaseEstimator, TransformerMixin):
```
I think a more natural place for this would be in `moabb/pipeline`

sylvchev replied:
Right, it is moved.

alexandrebarachant commented on examples/plot_cross_subject_ssvep.py:
```diff
@@ -0,0 +1,209 @@
+# proposed in Bin et al. (2009) https://doi.org/10.1088/1741-2560/6/4/046002
+
+
+class SSVEP_CCA(BaseEstimator, ClassifierMixin):
```
idem

sylvchev replied:
Ok, create a `pipelines/classifier.py`. I'll add some unit tests.
